Assignment 3

Statement of use of AI:
In Task 0, I used it for some guidance in setting up Ollama. In task 1, I used it to help me debug the problems I was having. In task 2, I used it where required in part c.


Task 0

In [66]:
# Part a 

# Set up ollama using commander prompt and the following is connecting to it using Ollama's API
import requests
import json

url = "http://localhost:11434/api/chat"

data = {
    "model": "gemma3:270m",
    "messages": [
        {
            "role": "user",
            "content": "Hello!"
        }
    ],
    "stream": False 
}

response = requests.post(url, json=data)

print(response.json()["message"]["content"])

Hello! How can I help you today?



In [67]:
# Part b
# In command prompt I loaded the 270m parameter verison.
url = "http://localhost:11434/api/chat"

data = {
    "model": "gemma3:270m",
    "messages": [
        {
            "role": "user",
            "content": "How many countires do pigeons live in?"
        }
    ],
    "stream": False 
}

response = requests.post(url, json=data)

print(response.json()["message"]["content"])

Pigeons live in a relatively small number of colonies, typically around 100-200.


In [68]:
# Part C
url = "http://localhost:11434/api/chat"

data = {
    "model": "gemma3:1b",
    "messages": [
        {
            "role": "user",
            "content": "Can penguins fly?"
        }
    ],
    "stream": False 
}

response = requests.post(url, json=data)

print(response.json()["message"]["content"])

No, penguins cannot fly. Despite their waddling gait and adorable appearance, they are flightless birds! 

They've evolved over millions of years to be incredibly well-suited for swimming and diving in the cold, icy waters of the Southern Hemisphere. 

Here's a breakdown of why they can't fly:

* **Wings:** Penguins have wings, but they're adapted for swimming, not flying. Their wings are shaped like flippers, perfect for propelling them through the water.
* **Weight and Body Structure:** They are heavy and have a compact body shape, making it difficult to generate the lift needed for flight.
* **Bone Structure:** Their bones are lightweight and hollow, crucial for buoyancy in the water.

**However, they *do* have some amazing adaptations for swimming:**

* **Powerful Flippers:** Their flippers are incredibly strong and maneuverable, allowing them to swim with exceptional speed and agility.
* **Dense Bones:** Their bones are dense and lightweight.

Do you have any other questions about

Task 1

In [74]:
# Task a
# Creating the function

Model_name = "gemma3:270m"

def classify_email(headline):
    url = "http://localhost:11434/api/generate"

    response = requests.post(
        url,
        json={
            "model": Model_name,
            "prompt": f"""
Classify the email headline into one of the following categories:
- spam
- work
- unknown

Respond with only one word. Do not explain. It must be one of the above categories. Only respond with either spam, work or unknown, do not respond with anything else.

Headline: {headline}
""",
            "stream": False 
         }
    )

    return response.json()["response"].strip().lower()


In [75]:
# Part b
with open("C:/Users/maisy/emails.csv", "r", encoding="utf-8") as f:
    headlines = f.read().splitlines()[1:]

for email in headlines:
    result = classify_email(email)
    print(result)

unknown
work
unknown
unknown
unknown
unknown
work
unknown
unknown
work
work
unknown


In [76]:
# Part c
Model_name= "gemma3:1b"

with open("C:/Users/maisy/emails.csv", "r", encoding="utf-8") as f:
    headlines = f.read().splitlines()[1:]

for email in headlines:
    result = classify_email(email)
    print(result)


spam
spam
spam
work
work
work
work
work
unknown
unknown
unknown
unknown


Comment: The 720m model has more unknowns with 7, compared to just 4 on the 1b version. The 720m version didn't recognise any as spam whereas the 1b version recognised 3 as spam. The recongised a similar amount as work, however the ones they recognised as work where different. They recognised a few the same but mostly different.

In [79]:
# Part c
import pandas as pd

# Running the models 3 times
results_270m = []
results_1b = []

for i in range(3):
    run_270m = []
    run_1b = []
   
    for email in headlines:
        Model_name = "gemma3:270m"
        run_270m.append(classify_email(email))

        Model_name = "gemma3:1b"
        run_1b.append(classify_email(email))

    results_270m.append(run_270m)
    results_1b.append(run_1b)

# Load into dataframe

df=pd.DataFrame({
    "headline": headlines,

    "270m_run1" : results_270m[0],
    "270m_run2" : results_270m[1],
    "270m_run3" : results_270m[2],

    "1b_run1": results_1b[0],
    "1b_run2": results_1b[1],
    "1b_run3": results_1b[2],
})

display(df)

,headline,270m_run1,270m_run2,270m_run3,1b_run1,1b_run2,1b_run3
0,URGENT: Your account will be suspended within ...,unknown,work,unknown,spam,spam,spam
1,Congratulations! You have won a 1000€ gift car...,work,spam,work,spam,spam,spam
2,Hot singles in your area are waiting to meet y...,unknown,work,unknown,spam,spam,spam
3,Re: Inheritance transfer of 4.5M USD pending y...,unknown,work,unknown,work,work,work
4,Meeting agenda for Thursday's project review,unknown,work,work,work,work,work
5,"Q3 budget report attached, please review by Fr...",unknown,unknown,unknown,work,work,work
6,Reminder: Annual performance review scheduled ...,work,work,work,work,work,work
7,"Updated draft of the manuscript, comments welcome",work,unknown,work,work,work,work
8,Quick question about last week,unknown,unknown,unknown,unknown,unknown,unknown
9,Following up,unknown,unknown,unknown,unknown,unknown,unknown


Comment: In the 1b version, only 1 categorisation changed in the three attempts in run 2, however it was changed back to unknown in run 3. However, in 270m version, there was a lot more changes over the three runs. This happened because smaller language models (270m) are less consistent and have a weaker understanding of context. Larger language models (1b) have more parameters and therefore usually produce more stable and accurate classifications.

Task 2

In [80]:
# Part a
import json

def classify_news(headline):
    url = "http://localhost:11434/api/generate"

    model_name = "gemma3:1b"   # used 1b as 4b was too slow
    response = requests.post(
        url,
        json={
            "model": model_name,
            "prompt": f"""

Classify the following financial news headlines.

Choose one topic from:
- earnings
- mergers
- regulation
- macroeconomics

Respond with only one word. Do not explain. It must be one of the above categories. Only respond with either earnings, mergers, regulation, macroeconomics, do not respond with anything else.

Choose one sentiment from: 
- positive
- negative
- neutral

Respond with only one word. Do not explain. It must be one of the above categories. Only respond with either positive, negative, netural, do not respond with anything else.

Return the result only in valid JSON format like this:
{{"topic":"earnings", "sentiment": "positive"}}

Headline: {headline}
""",
            "stream": False 
         }
    )

    result = response.json()["response"].strip()
    return json.loads(result)
    

In [83]:
# Part b

import pandas as pd

# Loading the dataset
with open("C:/Users/maisy/news.csv", "r", encoding="utf-8") as f:
    headlines = f.read().splitlines()[1:]

topics = []
sentiments = []

# Classifying the headlines
for headline in headlines:
    result = classify_news(headline)

    topics.append(result["topic"])
    sentiments.append(result["sentiment"])

# Add the results to a dataframe
df = pd.DataFrame({
    "headline": headlines,
    "topic": topics,
    "sentiment": sentiments
})

display(df)

,headline,topic,sentiment
0,Nordion Industries beats Q1 earnings estimates...,earnings,positive
1,Helvora Pharmaceuticals misses earnings foreca...,earnings,negative
2,"Aurelis Bank reports steady quarterly profit, ...",earnings,positive
3,Veridyne Logistics to acquire rival Trantec in...,mergers,positive
4,Antitrust regulators block proposed merger bet...,regulation,negative
5,Kestrel Semiconductor confirms early-stage mer...,mergers,positive
6,New EU AI Act compliance rules expected to rai...,regulation,negative
7,Finnish FSA grants Norvik Capital expanded lic...,mergers,positive
8,"Eurozone inflation cools to 2.1%, easing press...",earnings,positive
9,Rising interest rates weigh on Tessaro Real Es...,macroeconomics,negative


Part c

I gave Claude the dataset and the prompt and these are the results:

Headline: Nordion Industries beats Q1 earnings estimates as cloud revenue surges 28%
Topic: Earnings
Sentiment: Positive

Headline: Helvora Pharmaceuticals misses earnings forecast amid weak generics demand
Topic: Earnings
Sentiment: Negative

Headline: Aurelis Bank reports steady quarterly profit, in line with analyst expectations
Topic: Earnings
Sentiment: Neutral

Headline: Veridyne Logistics to acquire rival Trantec in 4.2 billion euro deal
Topic: Mergers & Acquisitions
Sentiment: Neutral

Headline: Antitrust regulators block proposed merger between Solenta and Marvex Energy
Topic: Mergers & Acquisitions
Sentiment: Negative

Headline: Kestrel Semiconductor confirms early-stage merger talks with Aldenfeld AG
Topic: Mergers & Acquisitions
Sentiment: Neutral

Headline: New EU AI Act compliance rules expected to raise costs for Lumavex by 12%
Topic: Regulation
Sentiment: Negative

Headline: Finnish FSA grants Norvik Capital expanded licence for cross-border operations
Topic: Regulation
Sentiment: Positive

Headline: Eurozone inflation cools to 2.1%, easing pressure on Drava Holdings borrowing costs
Topic: Macroeconomics
Sentiment: Positive

Headline: Rising interest rates weigh on Tessaro Real Estate as financing costs climb
Topic: Macroeconomics
Sentiment: Negative

Comparison:
For topic, they classified seven the same but three different. When they disagreed, I believe Claude to have accurately classified them while ollama did not. For example in the headline begining antitrust, I believe it to be about mergers, which Claude identified however, ollama classified it as regualtion. For sentiment, ollama never described it as netural, whereas Claude did 3 times, in these instances, I beleive them to be netural. In terms of positive and negative, they classifed it the same. Claude seemed to have more nuance describing some as neutral whereas ollama described them as just positive or negative. Overall, Claude is more accurate and this is because it is a larger and more heavily trained model meaning it can follow instructions more reliably and handle ambiugity and sentiment analysis more consistently.